javaslang/src/main/java/javaslang/Lazy.java

```diff
@@ -135,14 +133,11 @@ private Lazy(Supplier<? extends T> supplier) {
      */
     @Override
     public T get() {
-        // using a local var speeds up the double-check idiom by 25%, see Effective Java, Item 71
-        Supplier<? extends T> tmp = supplier;
-        if (tmp != null) {
+        if ((value == null) && (supplier != null)) {
```

Comment:
@paplorinc when I understood @shipilev right, the non-volatile read of `value`

```java
if (value == null && ...)
```

does not guarantee that (in the case of `value != null`) the second read does evaluate the same value:

```java
return value;
```

In other words your version is not deterministic and may produce unexpected results.

Am I right right?